src/v/cloud_storage/inventory/inv_consumer.cc

```diff
@@ -0,0 +1,259 @@
+        _total_size += sizeof(hash);
+    }
+}
+
+ss::future<>
+inventory_consumer::flush(inventory_consumer::write_all_t write_all_entries) {
+    auto h = _gate.hold();
+    vlog(
+      cst_log.trace,
+      "Flushing to disk, write_all_entries: {}",
+      write_all_entries);
+
+    _num_flushes += 1;
+    if (write_all_entries) {
+        for (auto& [ntp, flush_state] : _ntp_flush_states) {
+            if (!flush_state.hashes.empty()) {
+                co_await flush_ntp_hashes(
+                  ntp,
+                  std::move(flush_state.hashes),
+                  flush_state.next_file_name);
+            }
+        }
+        co_return;
+    }
+
+    // Set up a vector to sort entries. The ntp and next filename are not moved
+    // out of the map, only the hashes are moved out.
+    fragmented_vector<flush_entry> entries;
+    ranges::transform(
+      _ntp_flush_states, std::back_inserter(entries), [](auto& e) {
+          return flush_entry{
+            e.first, std::move(e.second.hashes), e.second.next_file_name};
+      });
+
+    ranges::sort(entries, ranges::greater(), [](const auto& e) {
+        return e.hashes.size();
+    });
+
+    const auto target_size = _max_hash_size_in_memory / 3;
+    vlog(
+      cst_log.trace,
+      "Before flush, size in memory {}, target {}",
+      _total_size,
+      target_size);
+
+    size_t idx = 0;
+    while (idx < entries.size() && _total_size >= target_size) {
+        _total_size -= sizeof(entries[idx].hashes[0])
+                       * entries[idx].hashes.size();
+
+        const auto path = entries[idx].ntp.path();
+
+        // Increment the next filename in the map. This must be done before
+        // moving the ntp. Even if the flush fails we have gaps in file names,
+        // which is fine because the filenames can be arbitrary and the hash
+        // loader will read all files in the NTP dir.
+        _ntp_flush_states[entries[idx].ntp].next_file_name += 1;
+
+        co_await flush_ntp_hashes(
+          std::move(entries[idx].ntp),
+          std::move(entries[idx].hashes),
+          entries[idx].file_name);
+
+        vlog(
+          cst_log.trace,
+          "Flushed ntp {}, size in memory {}, target {}",
+          path,
+          _total_size,
+          target_size);
+        idx += 1;
+    }
+
+    vlog(cst_log.trace, "After flush, size in memory {}", _total_size);
+
+    // Put back in all the hashes which were not flushed.
+    for (; idx < entries.size(); ++idx) {
+        _ntp_flush_states[entries[idx].ntp].hashes = std::move(
+          entries[idx].hashes);
+    }
+}
+
+ss::future<> inventory_consumer::flush_ntp_hashes(
+  model::ntp ntp, fragmented_vector<uint64_t> hashes, uint64_t file_name) {
+    auto ntp_hash_path = _hash_store_path / std::string_view{ntp.path()};
+    const auto ntp_hash_dir = ntp_hash_path.string();
+
+    if (!co_await ss::file_exists(ntp_hash_dir)) {
+        co_await ss::recursive_touch_directory(ntp_hash_dir);
+    }
```

Comment:
Is the expected usage that in between building inventory consumers, we'd delete the hash dirs?

Also is the hash dir going to be a subdirectory of the cloud cache? Still wondering if/how this will need to interact with space management and cloud cache trims

Reply:
Yes, the cleanup is implemented here https://github.com/redpanda-data/redpanda/pull/21386/files#diff-c85e68cc18d6fd68f542ecf14de615ca9ccaaab0917598efb09a08177d03a625R101-R112

The hash directory is expected to be outside the cache. This is a challenge for space management, but storing the data in cache could result in these hash files getting evicted before they are read since they won't be read as frequently. There was a discussion offline about this in the stand up and the end result of that discussion was that while keeping this data outside the cache will take up extra space, it does need to be available for scrubbers.

In terms of numbers, we see around 800MiB file size for hashing 100 million entries, so with something like 1000 partitions with this many entries each, we could see quite large space usage, close to 800GiB. 

A possible solution for these extreme cases (other than just making sure that much space is free) might be staggered scrubbing ie scrubbing a subset of NTPs at a time, which is something I am looking into.